pkg/skaffold/kubernetes/debugging/container_manager.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package debugging
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/event"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+)
+
+var (
+	// For testing
+	aggregatePodWatcher = kubernetes.AggregatePodWatcher
+
+	notifyDebuggingContainerStarted    = event.DebuggingContainerStarted
+	notifyDebuggingContainerTerminated = event.DebuggingContainerTerminated
+)
+
+type ContainerManager struct {
+	output      io.Writer
+	cli         *kubectl.CLI
+	podSelector kubernetes.PodSelector
+	namespaces  []string
+	active      map[string]string // set of containers that have been notified
+	aggregate   chan watch.Event
+}
+
+func NewContainerManager(out io.Writer, cli *kubectl.CLI, podSelector kubernetes.PodSelector, namespaces []string) *ContainerManager {
+	// Create the channel here as Stop() may be called before Start() when a build fails, thus
+	// avoiding the possibility of closing a nil channel. Channels are cheap.
+	return &ContainerManager{output: out, cli: cli, podSelector: podSelector, namespaces: namespaces, active: map[string]string{}, aggregate: make(chan watch.Event)}
+}
+
+func (d *ContainerManager) Start(ctx context.Context) error {
+	if d == nil {
```

Comment:
yeah this would be the 3rd podWatcher based control loop in skaffold with the exact same control flow. 

I would think a single podWatcher would be less resource intensive (the same "event messages" only need to be sent once for skaffold) on the API server side and on the skaffold side as well. 

I think, in a separate PR we should refactor first to a "visitor" pattern on a centralized podWatcher and make the logs, portforwards and debug listeners "register" with the podWatcher. 

In this visitor framework the implementation of the debug listener only would require line 94-99.

Reply:
OK. As discussed, we can push this now and refactor in a follow-up PR.